demo-circuit/src/common/data_structures.rs

```diff
@@ -0,0 +1,81 @@
+use cctp_primitives::{
+    type_mapping::FieldElement,
+    utils::{
+        data_structures::BackwardTransfer, get_bt_merkle_root,
+    },
+};
+
+#[derive(Clone, Debug)]
+/// The content of a withdrawal certificate
+pub struct WithdrawalCertificateData {
+    pub ledger_id: FieldElement,
+    pub epoch_id: u32,
+    /// Merkle root hash of all BTs from the certificate
+    pub bt_root: FieldElement,
+    pub quality: u64,
+    /// Reference to the state of the mainchain-to-sidechain transaction history.
+    /// Declares to which extent the sidechain processed forward transactions.
+    pub mcb_sc_txs_com: FieldElement,
+    pub ft_min_amount: u64,
+    pub btr_min_fee: u64,
+    /// Carries the reference to the sidechain state. (Currently the reference is
+    /// split over two field elements)
+    pub custom_fields: Vec<FieldElement>,
+}
+
+impl WithdrawalCertificateData {
+    /// Creates a Withdrawal Certificate that can be used for generating/verifying proofs
+    pub fn new(
+        ledger_id: FieldElement,
+        epoch_id: u32,
+        bt_list: Vec<BackwardTransfer>,
+        quality: u64,
+        mcb_sc_txs_com: FieldElement,
+        ft_min_amount: u64,
+        btr_min_fee: u64,
+        custom_fields: Vec<FieldElement>,
+    ) -> Self {
+        Self {
+            ledger_id,
+            epoch_id,
+            bt_root: get_bt_merkle_root(if bt_list.is_empty() {
+                None
+            } else {
+                Some(&bt_list)
+            })
+                .unwrap(),
+            quality,
+            mcb_sc_txs_com,
+            ft_min_amount,
+            btr_min_fee,
+            custom_fields,
+        }
+    }
+
+    /// Creates a Withdrawal Certificate that can be used for generating signatures
+    pub fn new_for_signing(
```

Comment:
Maybe this constructor is not needed Rust side: we receive a certificate without quality from the SDK, we compute msg_to_sign and number of valid sigs, then we have everything needed to instantiate a proper withdrawal certificate

Reply:
Maybe, it is used by any other library?

Reply:
I don't think so, it's a new change actually. We can remove it (unless you think you need it)

Reply:
If its not used anywhere but by java then I'll just remove it.

Reply:
(Y)